app/src/main/cpp/skyline/services/codec/IHardwareOpusDecoder.h

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2021 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#pragma once
+
+#include <opus.h>
+
+#include <common.h>
+#include <services/base_service.h>
+#include <kernel/types/KTransferMemory.h>
+
+namespace skyline::service::codec {
+    /**
+     * @brief This structure is located at the beginning of OpusDataIn for DecodeInterleaved*, with the actual opus packet following this
+     * @note These fields are big-endian
+     * @url https://github.com/switchbrew/libnx/blob/c5a9a909a91657a9818a3b7e18c9b91ff0cbb6e3/nx/include/switch/services/hwopus.h#L19
+     */
+    struct OpusDataHeader {
+        u32 sizeBe; //!< Size of the packet following this header, encoded in big-endian
+        u32 finalRangeBe; //!< Indicates the final range of the codec encoder's entropy coder, this can be left at zero, encoded in big-endian
+
+        u32 GetPacketSize() {
+            return util::SwapEndianness(sizeBe);
+        }
+    };
+    static_assert(sizeof(OpusDataHeader) == 0x8);
+
+    /**
+     * @return The required output buffer size for decoding an Opus stream with the given parameters
+     */
+    size_t CalculateOutBufferSize(i32 sampleRate, i32 channelCount, i32 frameSize);
+
+    static constexpr u32 OpusFullbandSampleRate{48000};
+    static constexpr u32 MaxFrameSizeNormal = OpusFullbandSampleRate * 0.040; //!< 40ms frame size limit for normal decoders
+    static constexpr u32 MaxFrameSizeMultiEx = OpusFullbandSampleRate * 0.120; //!< 120ms frame size limit for ex decoders added in 12.0.0
+    static constexpr u32 MaxInputBufferSize{0x600}; //!< Maximum allocated size of the input buffer
+
+    /**
+     * @brief This service decodes an Opus audio stream. Always created with OpenHardwareOpusDecoder by IHardwareOpusDecoderManager
+     * @url https://switchbrew.org/wiki/Audio_services#IHardwareOpusDecoder
+     */
+    class IHardwareOpusDecoder : public BaseService {
+      private:
+        std::shared_ptr<kernel::type::KTransferMemory> workBuffer;
+        OpusDecoder *decoderState{};
+        i32 sampleRate;
+        i32 channelCount;
+        i32 decoderOutputBufferSize;
+
+        /**
+         * @brief Resets the codec state to be equivalent to a freshly initialized state
+         */
+        void ResetContext();
+
+        /**
+         * @brief Called by all DecodeInterleaved* functions
+         * @param performanceInfo Whether to return decode time taken or not
```

Comment:
`decode time taken` isn't really valid English, `Whether to write the time taken to decode or not` would be more fitting but you could eliminate the comment entirely with a better variable name like `writeDecodeTime` since it's more explicit about what's being done while being the same length